### PR TITLE
Add support for tests with parentheses at the end of the name

### PR DESCRIPTION
**Pull request by jean-andre-gauthier**
* Add support for tests with parentheses at the end of the name by modifying the regex in `BaseJunitTestFrameworkStrategy.addPotentiallyParameterizedSuffixed()`, to only remove `(...)`, or `(...)[...]`, without matching an arbitrary number of parentheses / brackets, as was the case up until now
* Removes `AbstractPluginFuncTest.testLanguage()` and modifies the signature of `AbstractPluginFuncTest.writeTestSource`, to allow creating kotlin sample test classes
* Adds `ParenthesesFuncTest`, that tests https://github.com/gradle/test-retry-gradle-plugin/issues/101, https://github.com/gradle/test-retry-gradle-plugin/issues/123, and https://github.com/gradle/test-retry-gradle-plugin/issues/263

### CODE REVIEW

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/BaseJunitTestFrameworkStrategy.java [hunk 1]
```diff
@@ -118,7 +118,7 @@ private boolean processSpockTest(TestFilterBuilder filters, TestsReader testsRea
 
     private void addPotentiallyParameterizedSuffixed(TestFilterBuilder filters, String className, String name) {
         // It's a common pattern to add all the parameters on the end of a literal method name with []
-        String strippedParameterName = name.replaceAll("(?:\\([^)]*?\\)|\\[[^]]*?])*$", "");
+        String strippedParameterName = name.replaceAll("(?:\\([^)]*?\\))?(?:\\[[^]]*?](?:\\([^)]*?\\))?)?$", "");
```
Comment (snoopcheri) [inline plugin/src/main/java/org/gradle/testretry/internal/executer/framework/BaseJunitTestFrameworkStrategy.java:121]: 🤯

Comment (pshevche) [reply plugin/src/main/java/org/gradle/testretry/internal/executer/framework/BaseJunitTestFrameworkStrategy.java:121]: +1, can we add a comment explaining what those things should match? Or maybe a unit test?

**Review by pshevche — APPROVED** [review]

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/BaseJunitTestFrameworkStrategy.java (hunk 1 above)
Comment (jean-andre-gauthier) [reply plugin/src/main/java/org/gradle/testretry/internal/executer/framework/BaseJunitTestFrameworkStrategy.java]: I've realised I made the regex more complicated than necessary. I've updated it to match only `(...)`, or `(...)[...]`, added a comment, and an additional test case in `ParenthesesFuncTest`

File: plugin/src/test/groovy/org/gradle/testretry/ParenthesesFuncTest.groovy [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        expect:
+        def result = gradleRunner(gradleVersion, "test").build()
+        result.task(":test").outcome == TaskOutcome.SUCCESS
+
+        where:
+        [gradleVersion, scenarios] << [
+            // Kotlin plugin compatible from 6.8 onwards
+            GRADLE_VERSIONS_UNDER_TEST.findAll { GradleVersion.version(it) >= GradleVersion.version("6.8") },
+            [
+                new Tuple2<>({ bf -> setupJunit5Test(bf) }, junit5TestWithParentheses()),
```
Comment (leonard84) [inline plugin/src/test/groovy/org/gradle/testretry/ParenthesesFuncTest.groovy:45]:
```suggestion
                Tuple.tuple({ bf -> setupJunit5Test(bf) }, junit5TestWithParentheses()),
```
maybe with static import `groovy.lang.Tuple.tuple`

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/BaseJunitTestFrameworkStrategy.java [hunk 3]
```diff
@@ -118,7 +118,11 @@ private boolean processSpockTest(TestFilterBuilder filters, TestsReader testsRea
 
     private void addPotentiallyParameterizedSuffixed(TestFilterBuilder filters, String className, String name) {
         // It's a common pattern to add all the parameters on the end of a literal method name with []
-        String strippedParameterName = name.replaceAll("(?:\\([^)]*?\\)|\\[[^]]*?])*$", "");
+        // The regex takes care of removing trailing (...) or (...)[...], for e.g. the following cases
+        // * `test that contains (parentheses)()`
+        // * `test that contains (parentheses)(int, int)[1]`
+        // * `test(1, true) [0]`
+        String strippedParameterName = name.replaceAll("(?:\\([^)]*?\\))?(?:\\[[^]]*?])?$", "");
```
[the remark this answers is not in the document]
Comment (jean-andre-gauthier) [reply plugin/src/main/java/org/gradle/testretry/internal/executer/framework/BaseJunitTestFrameworkStrategy.java]: Done in 4b280bd

File: plugin/src/test/groovy/org/gradle/testretry/ParenthesesFuncTest.groovy (hunk 2 above)
Comment (jean-andre-gauthier) [reply plugin/src/test/groovy/org/gradle/testretry/ParenthesesFuncTest.groovy]: Done in 4b280bd

**Pull request merged**